Created depth-first-search.md (#2594)
Codecademy/docs: content/ai/concepts/search-algorithms/terms/depth-first-search/depth-first-search.md

---
Title: 'Depth-First Search'
Description: 'A shortest-path and traversing algorithm for unweighted graphs.'
Subjects:
  - 'Code Foundations'
  - 'Computer Science'
Tags:
  - 'Graphs'
  - 'Graph Search'
  - 'Algorithms'
CatalogContent:
  - 'learn-python'
  - 'paths/computer-science'
---

*Depth-first search (BFS)* is a traversing algorithm for unweighted graphs. This is a foundational algorithm in [graph](https://www.codecademy.com/resources/docs/general/graph) theory from which many other algorithms start.

## Features

Some of the features and constraints that define the use and functionality of a depth-first search algorithm include the following:

- A relatively simple and efficient algorithm for finding the shortest-path in graphs that do not have edge weights.

> - It has a time complexity of *O(|V| + |E|)*, which is the sum of the vertices and edges.

Reviewer: ```suggestion
- It has a time complexity of **O(|V| + |E|)**, which is the sum of the vertices and edges.
```

- It finds a solution but may or may not be optimal one.

## Implementation

The example below walks through a basic implementation of DFS that will take a starting node, an end node, and a graph as arguments. However, to start, a brief overview of the necessary steps or pseudocode will serve as an outline for the final implementation.
The depth-first search (DFS) algorithm starts with the initial node of graph G and goes deeper until we find the goal node or the node with no children.
Because of the recursive nature, stack data structure can be used to implement the DFS algorithm

The implementation below can be broken down into the following steps:

- First, create a stack with the total number of vertices in the graph.
- Now, choose any vertex as the starting point of traversal, and push that vertex into the stack.
- After that, push a non-visited vertex (adjacent to the vertex on the top of the stack) to the top of the stack.
- Now, repeat steps 3 and 4 until no vertices are left to visit from the vertex on the stack's top.
- If no vertex is left, go back and pop a vertex from the stack.
- Repeat steps 2, 3, and 4 until the stack is empty.
- 
## Example:

> ![image](https://github.com/Codecademy/docs/assets/100008872/b28fc472-761d-47e9-8189-a6b7865f7cb8)

Reviewer: Please follow the rules for uploading images from the style guide [here](https://github.com/Codecademy/docs/blob/main/documentation/style-guide.md#file-hosting).

> ## code:

Reviewer: ```suggestion
The following is an implementation in Python:

```

```python
#visited keepds record of visited nodes
def dfs(graph,start,goal,stack,visited):
    stack.append(start)
    visited.append(start)
    print('The path traversed is:')
    while stack:
        #pop from stack if visited 
        element=stack.pop()
        print(element,end=" ")
        if(element==goal):
            break
        for neighbor in graph[element]:
            if neighbor not in visited:
                stack.append(neighbor)
                visited.append(neighbor)
#dictionary for given graph 
graph={
    'A':['C','B'],
    'B':['E','D'],
    'C':['G','F'],
    'D':[],
    'E':['I','H'],
    'F':[],
    'G':['J'],
    'H':[],
    'I':[],
    'J':[]
}
start='A'
goal='J'
visited=[]
stack=[]

dfs(graph,start,goal,visited,stack)
```

> ## output:

Reviewer: ```suggestion
```
We don't use images for output that's just text to the console, those should be illustrated in a ` ```shell` block.
Alternatively you could make the example a python codebyte, which is runnable by the reader and doesn't need an output illustration.

Author: > We don't use images for output that's just text to the console, those should be illustrated in a `` ```shell `` block. Alternatively you could make the example a python codebyte, which is runnable by the reader and doesn't need an output illustration.

Done with all the changes respected sir @SSwiniarski please approve these changes ,done it as per rules.

> ![image](https://github.com/Codecademy/docs/assets/100008872/7b040a7d-90b4-4188-8ef0-067f0507695d)

Reviewer: Please follow the rules for uploading images from the style guide [here](https://github.com/Codecademy/docs/blob/main/documentation/style-guide.md#file-hosting).

---
Title: 'Depth-First Search'
Description: 'A shortest-path and traversing algorithm for unweighted graphs.'
Subjects:
  - 'Code Foundations'
  - 'Computer Science'
Tags:
  - 'Graphs'
  - 'Graph Search'
  - 'Algorithms'
CatalogContent:
  - 'learn-python'
  - 'paths/computer-science'
---

> *Depth-first search (DFS)* is a traversing algorithm for unweighted graphs. This is a foundational algorithm in [graph](https://www.codecademy.com/resources/docs/general/graph) theory from which many other algorithms start.

Reviewer 2: ```suggestion
**Depth-first search (DFS)** is a traversing algorithm for unweighted graphs. This is a foundational algorithm in [graph](https://www.codecademy.com/resources/docs/general/graph) theory from which many other algorithms start.
```

Reviewer 2: It looks like one of our existing entries, breadth-first search, was used as a starting point for this submission. There isn't anything wrong with using existing content as a reference, or inspiration, but a new submission cannot directly copy content from another. There are several instances in this entry where sentences from the previous entry have not been edited or rephrased for this submission. Please review and revise this entry to ensure that all of the content is original.

Author: > It looks like one of our existing entries, breadth-first search, was used as a starting point for this submission. There isn't anything wrong with using existing content as a reference, or inspiration, but a new submission cannot directly copy content from another. There are several instances in this entry where sentences from the previous entry have not been edited or rephrased for this submission. Please review and revise this entry to ensure that all of the content is original.

Done respected sir please review my PR.

---
Title: 'Depth-First Search'
Description: 'A traversing algorithm for unweighted graphs.'
Subjects:
  - 'Code Foundations'
  - 'Computer Science'

>   - 'Artificial Intellignce'

Reviewer 2: ```suggestion
```

Tags:
  - 'Graphs'
  - 'Graph Search'
  - 'Algorithms'
CatalogContent:

>   - 'learn-python'

Reviewer 2: ```suggestion
  - 'learn-python-3'
```

  - 'paths/computer-science'
---

> **Depth-first search (DFS)** is a traversing algorithm for unweighted graphs.Like BFS (Breadth First Search), it is a foundational algorithm in [graph](https://www.codecademy.com/resources/docs/general/graph) theory from which many other algorithms begin.

Reviewer 2: ```suggestion
**Depth-first search (DFS)** is a traversing algorithm for unweighted graphs.Like BFS (Breadth-first search), it is a foundational algorithm in [graph](https://www.codecademy.com/resources/docs/general/graph) theory from which many other algorithms begin.
```

## Features

> Some of the features and constraints that define the use and functionality of the DFS algorithm include the following:

Reviewer 2: ```suggestion
Some of the features and constraints that define the use and functionality of the DFS algorithm include the following:

```

> - DFS traverses a graph by following a given path as far as possible.Once a dead end is reached, the algorithm backtracks to explore the most recent unvisited neighbours.  A stack is used to remember how to get the next vertex to start a search, when a dead end is found in any iteration. 

Reviewer 2: ```suggestion
- DFS traverses a graph by following a given path as far as possible. Once a dead end is reached, the algorithm backtracks to explore the most recent unvisited neighbours.  A stack is used to retrieve the next vertex, to restart a search, when a dead end is found in any given iteration. 
```

> - It can be a relatively simple algorithm that can be built upon for finding cycles within a graph, to conduct a topological sort of vertices, as well as a range of other useful applications.

Reviewer 2: ```suggestion
- It is a relatively simple algorithm that can be built upon for finding cycles within a graph, to conduct a topological sort of vertices, as well as a range of other useful applications.
```

- It has a time complexity of **O(|V| + |E|)**, which is the sum of the vertices and edges.
- It finds a solution but may or may not be optimal one as in BFS.
- The difference between DFS and BFS is that DFS takes less time than BFS and finds a solution.
- The disadvantage of DFS is that it may get stuck in infinite loop if depth is infinite.

## Implementation

> The example below walks through a basic implementation of DFS that will take a starting node, an end node, and a graph as arguments. The algorithm will return the path it navigated between starting and end node provided.

Reviewer 2: ```suggestion
The example below walks through a basic implementation of DFS that will take a starting node, an end node, and a graph as arguments. The algorithm will return the path it navigated between starting and end node provided.

```

The depth-first search (DFS) algorithm starts with the initial node of graph G and goes deeper until we find the goal node or the node with no children.

> Because of the recursive nature of the traversal process, stack data structure can be used to implement the DFS algorithm.

Reviewer 2: ```suggestion
Because of the recursive nature of the traversal process, a stack data structure can be used to implement the DFS algorithm.
```

The implementation below can be broken down into the following steps:

> - First, create a stack with the total number of vertices in the graph.

Reviewer 2: ```suggestion
1. First, initiate a stack with the starting vertex for the traversal.
```

> - Now, choose any vertex as the starting point of traversal, and push that vertex into the stack.

Reviewer 2: ```suggestion
2. Pop from the stack and set this vertex as the "current" element or node.
3. Now, find the neighboring vertexes (of the current node), and if they haven't been visited push them into the stack.
```

> - After that, push a non-visited vertex (adjacent to the vertex on the top of the stack) to the top of the stack.

Reviewer 2: ```suggestion
```

> - Now, repeat steps 3 and 4 until no vertices are left to visit from the vertex on the stack's top.

Reviewer 2: ```suggestion
```

> - If no vertex is left, go back and pop a vertex from the stack.

Reviewer 2: ```suggestion
4. If no unvisited vertexes remain, go back and pop a vertex from the stack.
```

> - Repeat steps 2, 3, and 4 until the stack is empty.

Reviewer 2: ```suggestion
5. Repeat steps 2, 3, and 4 until the stack is empty.
```

> ## Example:

Reviewer 2: ```suggestion
## Example

```

https://raw.githubusercontent.com/Codecademy/docs/main/media/Depth-First-Search-Example.png

## code:
The following is an implementation in Python:

```python

> #visited keepds record of visited nodes

Reviewer 2: ```suggestion
# The visited variable keeps a record of the nodes explored
```

def dfs(graph,start,goal,stack,visited):
    stack.append(start)
    visited.append(start)
    print('The path traversed is:')
    while stack:

>         #pop from stack if visited 

Reviewer 2: ```suggestion
        # Pop from stack to set current element  
```

        element=stack.pop()
        print(element,end=" ")
        if(element==goal):
            break
        for neighbor in graph[element]:
            if neighbor not in visited:
                stack.append(neighbor)
                visited.append(neighbor)
#dictionary for given graph 
graph={
    'A':['C','B'],
    'B':['E','D'],
    'C':['G','F'],
    'D':[],
    'E':['I','H'],
    'F':[],
    'G':['J'],
    'H':[],
    'I':[],
    'J':[]
}
start='A'
goal='J'

> visited=[]

Author: > seems a little odd to me that visited and stack are passed as arguments.

Applied my logic here which will be easy for all to understand. Also gave names accordingly so that its easy to understand.

stack=[]

dfs(graph,start,goal,visited,stack)
```

```shell
The path traversed is:

> A B D E H I C F G J >

Reviewer 2: ```suggestion
A B D E H I C F G J 
```
